context: debut

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
context: debut

rockwotj commented on src/v/context/context.h:
```diff
@@ -0,0 +1,477 @@
+
+} // namespace context
+
+class context_frame {
```
why call it a “frame”?

nvartolomei replied:
https://github.com/redpanda-data/redpanda/pull/29083#discussion_r2637517583

nvartolomei commented:
Because they look similar to stack frames, logically not physically. In the
way they form a parent child relationship and the way they store
information.

Then “context” at a particular stack frame is all the ancestor context
frames taken together. So that’s why it’s not called just context. In the
current iteration, all the information is stored in a particular context so
that might end up unnecessary. Still experimenting.

On Sun, 21 Dec 2025 at 05:22, Tyler Rockwood ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/context/context.h
> <https://github.com/redpanda-data/redpanda/pull/29083#discussion_r2637517583>
> :
>
> > +    cancel_handle(const cancel_handle&) noexcept;
> +    cancel_handle(cancel_handle&&) noexcept;
> +    cancel_handle& operator=(const cancel_handle&) noexcept;
> +    cancel_handle& operator=(cancel_handle&&) noexcept;
> +    ~cancel_handle();
> +
> +    /// \brief Cancels the associated frame and propagates to descendants.
> +    void trigger(cancel_cause cause) noexcept;
> +
> +private:
> +    context_frame* frame_;
> +};
> +
> +} // namespace context
> +
> +class context_frame {
>
> why call it a “frame”?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/29083#pullrequestreview-3601447284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWLXBYCKHAV2LBCV3CD4CYG6BAVCNFSM6AAAAACPUKLQG6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTMMBRGQ2DOMRYGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nvartolomei committed:
context: add linker mixin

Allows a context frame to depend on multiple parent contexts, propagating
state from any linked parent.

rockwotj commented on src/v/context/context.h:
```diff
@@ -0,0 +1,510 @@
+/// Low resolution system clock for estimating wall-clock deadlines.
+using lowres_system_clock = seastar::lowres_system_clock;
+
+/// Infinite future (no deadline).
```
Please tell me we're not going to have overflow issues :)

nvartolomei replied:
If our reference clock is in nanoseconds, can it overflow?

I did not think much about it though before you asking.

rockwotj replied:
I am more worried about calling time left, and places where you do now + time left.

This is why I like absl's time and duration classes because they use saturating arithmetic so this isn't an issue. maybe we need our own duration types that are not prone to this...

nvartolomei replied:
Oh :(

Yeah. Would be nice to get rid of this footgun. Any suggestions?

Removing time_left() is my first option.

rockwotj replied:
> Any suggestions?

https://github.com/redpanda-data/redpanda/pull/29123 (I've been meaning to do this for a while)

rockwotj commented on src/v/context/context_frame.h:
````diff
@@ -0,0 +1,253 @@
+// by the Apache License, Version 2.0
+
+/// \file context_frame.h
+/// \brief Mixin-composable context frame.
+///
+/// This header provides `context_frame<Mixins...>` for composing contexts with
+/// zero-cost, opt-in features. Most users only need `context/context.h` for
+/// `context_ref` and `background()`.
+///
+/// ## Usage
+///
+/// ```cpp
+/// #include "context/context_frame.h"
+/// #include "context/deadline_timer.h"
+///
+/// using my_frame = context::context_frame<context::deadline_timer>;
+/// my_frame frame{parent, context::with<context::deadline_timer>(100ms)};
+/// ```
+///
+/// ## Built-in Mixins
+///
+/// - `deadline_timer` (deadline_timer.h): Automatic cancellation on timeout.
+/// - `abort_source` (abort_source.h): Bridge to Seastar's abort_source.
````
I think I would be happier if we `s/context_frame/context` because I really see all these as "contexts" and you get references to them. Because "context_ref" is really a reference to a frame not a context, so it should be called `context_frame_ref`? I think having a simple `context` and `context_ref` as the two main classes is simpler, but I have to learn this mixin thing more first.

My gut reaction is the mixin thing feels like it's a bit over complicated, but I can probably be convinced.

nvartolomei replied:
> My gut reaction is the mixin thing feels like it's a bit over complicated, but I can probably be convinced.

It is. For the end user. It also opens the door to binary bloat if multiple users start writing `context_frame<deadline_timer, abort_source>`, other writes `context_frame<abort_source, deadline_timer>`, another `context_frame<linked>`, etc.

That's the price you pay for zero-cost (no dynamic casts, no typeid checks, no vtable overhead).

What we should do is make raise a bit the bar for using this pattern and instead provide out-of-the-box few simpler abstractions that are built from these mixins so that each user does not have to built its own.

Still, compile times aren't cheap but during linking this will reduce binary bloat, I reckon. Can probably also just hide it behind a pimpl (with a dynamic allocation) and have it implicitly convert to context_ref as well.

rockwotj replied:
I don't know, are there really that many variations of these patterns that we need generic mixins or can we just spell out the 4 variations we need (base, abort source, deadline, deadline and abort source) and that will be good enough? They will all get type erased to context_ref and will be a virtual function call anyways

rockwotj reviewed (commented):
I am feeling this pain in CT (I tagged you the other review). When can we have something like this and also a compat story for `retry_chain_node` 😄

nvartolomei commented:
> I am feeling this pain in CT (I tagged you the other review). When can we have something like this and also a compat story for retry_chain_node 😄

_sketched on a napkin_

From rtc:

```cpp
using ctx_t = context_frame<deadline_timer>;

ctx_t ctx{todo(), with<deadline_timer>(rtc.get_deadline())}
auto sub = rtc.root_abort_source().root_abort_source().subscribe([ctx] mutable {
  ctx.trigger_cancel(cancel_cause::manual);
});
```

To rtc:

```cpp
using ctx_t = context_frame<abort_source>;
ctx_t ctx{todo()};

rtc retry_chain_node(ctx.as()); // deadlines and backoffs are optional
```

It sort of already works. Not sure we need a mixin here.

There is this question which I haven't come up with how to answer: When my additional data should be a mixin (i.e. kafka_request) or instead my context is a member of class kafka_request.

nvartolomei commented:
idea

```cpp
auto t = context::tenacity{ctx, /* per permit timeout, backoff strategy, max retries, etc */};

while (true) {
  auto permit = co_await t.permit(); //< sleeps (abortable by parent context ofc) for backoff + jitter; returns a std::expected<tenacity_try_context, permit_error_code>
  if (!permit) {
    // bail out out
  }
  auto http_result = client->request(permit, ...);
  // ...
}

```

alternatively (hides the loop, waiting on the permit, harder to pass the wrong context after copy/paste)

```cpp
auto result = co_await t.run([&client](auto permit) {
    return client->request(permit, ...);
});
```

this doesn't have to live in `src/v/context` at all and can be built outside

nvartolomei committed:
ssx: add context-aware sleep

Add ssx::sleep() that integrates with the context cancellation hierarchy.
Unlike seastar::sleep_abortable which requires an abort_source subscription
for each sleep call, this directly uses the context's cancellation mechanism,
avoiding an extra context frame with abort_source mixin.

Allocation comparison: seastar sleep allocates 192 bytes, while this
allocates 200 bytes with ref counting enabled.

The ssx::context_sleep_aborted exception extends seastar::sleep_aborted for
compatibility with existing catch handlers while preserving the cancel cause.

nvartolomei committed:
utils: add context-aware retry_loop_state utility

Add a retry_loop_state class that implements exponential backoff with
jitter for retry loops. The class integrates with the context system
to respect cancellation and deadlines, returning nullopt when the
context is cancelled or when attempts are exhausted.
